README.md
# apcupsd-mqtt-exporter

This is a docker application to publish apcupsd metrics to MQTT.

## Running

```bash
docker run --name apcupsd-mqtt-exporter \
  --env-file .env \
  --network host \
  ghcr.io/joeyeamigh/apcupsd-mqtt-exporter:latest
```

## Environment

```sh
APCUPSD_HOST="127.0.0.1"  # host running apcupsd
APCUPSD_PORT=3551         # port apcupsd is listening on
APCUPSD_STRIP_UNITS=true  # strip units from apcupsd values
APCUPSD_POLL_INTERVAL=10  # seconds between polling apcupsd
APCUPSD_POLL_TIMEOUT=5    # seconds between polling apcupsd
MQTT_HOST="127.0.0.1"     # host running MQTT broker
MQTT_PORT=1883            # port MQTT broker is listening on
MQTT_USERNAME=""          # MQTT username (optional)
MQTT_PASSWORD=""          # MQTT password (optional)
MQTT_CLIENT_ID="apcupsd"  # MQTT client ID
MQTT_TOPIC="apcupsd/ups"  # MQTT topic to publish to
MQTT_SUFFIX=""            # MQTT topic suffix (optional)
HOME_ASSISTANT_MODE=false # publish MQTT messages in Home Assistant-compatible JSON
```

To use locally with the `.env` file, make sure to enable the "dotenv" feature, e.g.:

```bash
cargo run -F dotenv
```

To use with Home Assistant, the following `.env` will enable automatic discovery:

```sh
APCUPSD_HOST="localhost"                     # host running apcupsd
APCUPSD_PORT=3551                            # port apcupsd is listening on
APCUPSD_STRIP_UNITS=true                     # strip units from apcupsd values
APCUPSD_POLL_INTERVAL=10                     # seconds between polling apcupsd
APCUPSD_POLL_TIMEOUT=5                       # seconds between polling apcupsd
MQTT_HOST="localhost"                        # host running MQTT broker
MQTT_PORT=1883                               # port MQTT broker is listening on
MQTT_USERNAME=""                             # MQTT username (optional)
MQTT_PASSWORD=""                             # MQTT password (optional)
MQTT_CLIENT_ID="upcupsd"                     # MQTT client ID
MQTT_TOPIC="homeassistant/sensor/ups"        # MQTT topic to publish to
MQTT_SUFFIX="status"                         # MQTT topic suffix (optional)
HOME_ASSISTANT_MODE=true                     # publish MQTT messages in Home Assistant-compatible JSON
HOME_ASSISTANT_UUID_PREFIX="apcupsd_"        # prefix for Home Assistant UUIDs
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.2.3] - 2023-11-8

Shrinking size of docker container from 867MB to 18.8MB.

## [0.2.2] - 2023-9-2

Changing HomeAssistant mode have a 5 min unavailable timeout.

## [0.2.1] - 2023-9-2

Better error message for timeout and image version bump.

## [0.2.0] - 2023-8-30

Added support for Home Assistant formatting and polling timeouts.

## [0.1.0] - 2023-8-27

First release! See README for features.


src/main.rs
#![feature(let_chains)]

use tracing::info;

mod apcupsd;
mod config;
mod mqtt;

#[tokio::main]
async fn main() {
  init_logger();

  let config = config::Config::new();

  let (mut mqtt_sender, rx, tx) = mqtt::MQTTSender::new(&config);

  info!("Starting apcupsd_mqtt_exporter");

  tokio::spawn(async move {
    apcupsd::APCUPSdPolling::new(&config, tx).poll().await;
  });

  mqtt_sender.listen(rx).await;
}

fn init_logger() {
  use tracing::metadata::LevelFilter;
  use tracing_subscriber::{
    fmt, prelude::__tracing_subscriber_SubscriberExt, util::SubscriberInitExt, EnvFilter, Layer,
  };

  #[cfg(debug_assertions)]
  let filter_directives = if let Ok(filter) = std::env::var("RUST_LOG") {
    filter
  } else {
    "apcupsd_mqtt_exporter=trace".to_string()
  };

  #[cfg(debug_assertions)]
  let filter = EnvFilter::builder()
    .with_default_directive(LevelFilter::TRACE.into())
    .parse_lossy(filter_directives);

  #[cfg(not(debug_assertions))]
  let filter_directives = if let Ok(filter) = std::env::var("RUST_LOG") {
    filter
  } else {
    "apcupsd_mqtt_exporter=info".to_string()
  };

  #[cfg(not(debug_assertions))]
  let filter = EnvFilter::builder()
    .with_default_directive(LevelFilter::TRACE.into())
    .parse_lossy(filter_directives);

  tracing_subscriber::registry()
    .with(fmt::layer().with_filter(filter))
    .init();
}


src/mqtt.rs
use super::config::Config;
use std::collections::HashMap;
use tokio::sync::mpsc;
use tracing::{debug, warn};

pub struct MQTTSender {
  client: rumqttc::AsyncClient,
  eventloop: rumqttc::EventLoop,
  topic: String,
  suffix: String,
  home_assistant_mode: bool,
  home_assistant_uuid_prefix: String,
}

type MQTTNewSenderReturn = (
  MQTTSender,
  mpsc::Receiver<HashMap<String, String>>,
  mpsc::Sender<HashMap<String, String>>,
);

impl MQTTSender {
  pub fn new(config: &Config) -> MQTTNewSenderReturn {
    let mut options = rumqttc::MqttOptions::new(
      config.mqtt_client_id.clone(),
      config.mqtt_host.clone(),
      config.mqtt_port,
    );

    if config.mqtt_has_credentials {
      options.set_credentials(config.mqtt_username.clone(), config.mqtt_password.clone());
    }

    let (client, eventloop) = rumqttc::AsyncClient::new(options, 1);

    let (tx, rx) = mpsc::channel(1);

    (
      Self {
        client,
        eventloop,
        topic: config.mqtt_topic.clone(),
        suffix: config.mqtt_suffix.clone(),
        home_assistant_mode: config.home_assistant_mode,
        home_assistant_uuid_prefix: config.home_assistant_uuid_prefix.clone(),
      },
      rx,
      tx,
    )
  }

  pub async fn listen(&mut self, mut rx: mpsc::Receiver<HashMap<String, String>>) {
    let qos = rumqttc::QoS::AtLeastOnce;
    let retain = false;
    let topic = self.topic.clone();
    let suffix = self.suffix.clone();
    let client = self.client.clone();
    let home_assistant_mode = self.home_assistant_mode;
    let home_assistant_uuid_prefix = self.home_assistant_uuid_prefix.clone();

    tokio::spawn(async move {
      let mut seen_before: HashMap<String, bool> = HashMap::new();

      loop {
        let data = rx.recv().await;
        let data = if let Some(data) = data {
          data
        } else {
          continue;
        };

        debug!("Received data from apcaccess: {:?}", data);

        if home_assistant_mode {
          for (param, value) in data.clone() {
            let uuid = format!("{}{}", home_assistant_uuid_prefix, param);

            let config_topic = format!("{}/{}/config", topic, uuid);
            let state_topic = format!("{}/{}/{}", topic, uuid, suffix);

            if !seen_before.contains_key(&param) {
              let payload = serde_json::json!({
                "name": format!("{}", param),
                "unique_id": uuid,
                "device": {
                  "identifiers": format!("{}", topic),
                  "name": "APC UPS",
                },
                "state_topic": state_topic,
                "value_template": "{{ value_json.state }}",
                "expire_after": 300,
              })
              .to_string();

              debug!(
                "Publishing Initial Home Assistant Config MQTT message: {} -> {:?}",
                config_topic, payload
              );

              if let Err(err) = client.publish(config_topic, qos, retain, payload).await {
                warn!(
                  "Failed to publish Initial Home Assistant Config MQTT message: {:?}",
                  err
                );
              }

              seen_before.insert(param.clone(), true);
            }

            let payload = serde_json::json!({ "state": value }).to_string();

            debug!(
              "Publishing Home Assistant State MQTT message: {} -> {:?}",
              state_topic, payload
            );

            if let Err(err) = client.publish(state_topic, qos, retain, payload).await {
              warn!("Failed to publish Home Assistant State MQTT message: {:?}", err);
            }
          }

          continue;
        }

        for (param, value) in data {
          let mut topic = format!("{}/{}", topic, param.to_lowercase().replace(' ', "-"));
          if !suffix.is_empty() {
            topic = format!("{}/{}", topic, suffix);
          }

          debug!("Publishing MQTT message: {} -> {:?}", topic, value.clone());

          if let Err(err) = client.publish(topic, qos, retain, value).await {
            warn!("Failed to publish MQTT message: {:?}", err);
          }
        }
      }
    });

    loop {
      let _ = self.eventloop.poll().await;
      // let mqtt_msg = self.eventloop.poll().await;
      // trace!("Received MQTT message: {:?}", mqtt_msg);
    }
  }
}


src/apcupsd.rs
use super::config;

use apcaccess::{APCAccess, APCAccessConfig};
use std::collections::HashMap;
use tokio::sync::mpsc;
use tracing::{error, trace};

pub struct APCUPSdPolling {
  apc: APCAccess,
  poll_interval: u64,
  timeout: u64,

  tx: mpsc::Sender<HashMap<String, String>>,
}

impl APCUPSdPolling {
  pub fn new(app_config: &config::Config, tx: mpsc::Sender<HashMap<String, String>>) -> Self {
    let config = APCAccessConfig {
      host: app_config.apcupsd_host.clone(),
      port: app_config.apcupsd_port,
      strip_units: app_config.apcupsd_strip_units,
      timeout: app_config.apcupsd_poll_timeout,
    };

    Self {
      apc: APCAccess::new(Some(config)),
      poll_interval: app_config.apcupsd_poll_interval,
      timeout: app_config.apcupsd_poll_timeout,

      tx,
    }
  }

  pub async fn poll(&self) {
    loop {
      trace!("polling apcupsd");
      let apc = self.apc.clone();

      let data = match tokio::time::timeout(
        std::time::Duration::from_secs(self.timeout),
        tokio::spawn(async move { apc.fetch() }),
      )
      .await
      {
        Ok(data) => match data {
          Ok(data) => match data {
            Ok(data) => data,
            Err(err) => {
              error!("failed to fetch data from apcupsd: {:?}", err);
              continue;
            }
          },
          Err(err) => {
            error!("failed to join tokio thread {:?}", err);
            continue;
          }
        },
        Err(_) => {
          error!(
            "timeout while fetching data from apcupsd: timeout of {} seconds elapsed",
            self.timeout
          );
          continue;
        }
      };

      trace!("successfully received data from apcupsd: {:?}", data);

      if let Err(err) = self.tx.send(data).await {
        error!("Failed to send data to MQTT: {:?}", err);
      }

      tokio::time::sleep(tokio::time::Duration::from_secs(self.poll_interval)).await;
    }
  }
}


src/config.rs
// APCUPSD_HOST: "127.0.0.1"         # host running apcupsd
// APCUPSD_PORT: 3551                # port apcupsd is listening on
// APCUPSD_STRIP_UNITS: true         # strip units from apcupsd values
// APCUPSD_POLL_INTERVAL: 10         # seconds between polling apcupsd
// APCUPSD_POLL_TIMEOUT: 5           # seconds to wait for apcupsd to respond
// MQTT_HOST: "127.0.0.1"            # host running MQTT broker
// MQTT_PORT: 1883                   # port MQTT broker is listening on
// MQTT_USERNAME: ""                 # MQTT username (optional)
// MQTT_PASSWORD: ""                 # MQTT password (optional)
// MQTT_CLIENT_ID: "apcupsd"         # MQTT client ID
// MQTT_TOPIC: "apcupsd/ups"         # MQTT topic to publish to
// MQTT_SUFFIX: ""                   # suffix to append to MQTT topic
// HOME_ASSISTANT_MODE: false        # publish MQTT messages in Home Assistant-compatible JSON
// HOME_ASSISTANT_UUID_PREFIX: ""    # prefix for Home Assistant UUIDs

use tracing::trace;

#[derive(Debug)]
pub struct Config {
  pub apcupsd_host: String,
  pub apcupsd_port: u16,
  pub apcupsd_strip_units: bool,
  pub apcupsd_poll_interval: u64,
  pub apcupsd_poll_timeout: u64,
  pub mqtt_host: String,
  pub mqtt_port: u16,
  pub mqtt_has_credentials: bool,
  pub mqtt_username: String,
  pub mqtt_password: String,
  pub mqtt_client_id: String,
  pub mqtt_topic: String,
  pub mqtt_suffix: String,
  pub home_assistant_mode: bool,
  pub home_assistant_uuid_prefix: String,
}

impl Default for Config {
  fn default() -> Self {
    Self {
      apcupsd_host: "127.0.0.1".to_string(),
      apcupsd_port: 3551,
      apcupsd_strip_units: true,
      apcupsd_poll_interval: 10,
      apcupsd_poll_timeout: 5,
      mqtt_host: "127.0.0.1".to_string(),
      mqtt_port: 1883,
      mqtt_has_credentials: false,
      mqtt_username: "".to_string(),
      mqtt_password: "".to_string(),
      mqtt_client_id: "apcupsd".to_string(),
      mqtt_topic: "apcupsd/ups".to_string(),
      mqtt_suffix: "".to_string(),
      home_assistant_mode: false,
      home_assistant_uuid_prefix: "".to_string(),
    }
  }
}

impl Config {
  pub fn new() -> Self {
    #[cfg(feature = "dotenv")]
    dotenv::dotenv().ok();

    let mut config = Config::default();

    if let Ok(apcupsd_host) = std::env::var("APCUPSD_HOST") {
      config.apcupsd_host = apcupsd_host;
    }

    if let Ok(apcupsd_port) = std::env::var("APCUPSD_PORT") {
      config.apcupsd_port = apcupsd_port.parse::<u16>().unwrap();
    }

    if let Ok(apcupsd_strip_units) = std::env::var("APCUPSD_STRIP_UNITS") {
      config.apcupsd_strip_units = apcupsd_strip_units.parse::<bool>().unwrap();
    }

    if let Ok(apcupsd_poll_interval) = std::env::var("APCUPSD_POLL_INTERVAL") {
      config.apcupsd_poll_interval = apcupsd_poll_interval.parse::<u64>().unwrap();
    }

    if let Ok(apcupsd_poll_timeout) = std::env::var("APCUPSD_POLL_TIMEOUT") {
      config.apcupsd_poll_timeout = apcupsd_poll_timeout.parse::<u64>().unwrap();
    }

    if let Ok(mqtt_host) = std::env::var("MQTT_HOST") {
      config.mqtt_host = mqtt_host;
    }

    if let Ok(mqtt_port) = std::env::var("MQTT_PORT") {
      config.mqtt_port = mqtt_port.parse::<u16>().unwrap();
    }

    if let Ok(mqtt_username) = std::env::var("MQTT_USERNAME")
        && let Ok(mqtt_password) = std::env::var("MQTT_PASSWORD") {
      config.mqtt_username = mqtt_username;
      config.mqtt_password = mqtt_password;
      config.mqtt_has_credentials = true;
    }

    if let Ok(mqtt_client_id) = std::env::var("MQTT_CLIENT_ID") {
      config.mqtt_client_id = mqtt_client_id;
    }

    if let Ok(mqtt_topic) = std::env::var("MQTT_TOPIC") {
      config.mqtt_topic = mqtt_topic;
    }

    if let Ok(mqtt_suffix) = std::env::var("MQTT_SUFFIX") {
      config.mqtt_suffix = mqtt_suffix;
    }

    if let Ok(home_assistant_mode) = std::env::var("HOME_ASSISTANT_MODE") {
      config.home_assistant_mode = home_assistant_mode.parse::<bool>().unwrap();
    }

    if let Ok(home_assistant_uuid_prefix) = std::env::var("HOME_ASSISTANT_UUID_PREFIX") {
      config.home_assistant_uuid_prefix = home_assistant_uuid_prefix;
    }

    if config.home_assistant_mode {
      if config.mqtt_suffix.is_empty() {
        config.mqtt_suffix = "status".to_string();
      }

      if config.home_assistant_uuid_prefix.is_empty() {
        config.home_assistant_uuid_prefix = "apcupsd_".to_string();
      }
    }

    trace!("initialized config: {:?}", config);

    config
  }
}
